### PR TITLE
2880 timebox create a poc to fetch emails from the gmail api

### LINKED ISSUE #2880: [Timebox] Create a POC to fetch emails from the Gmail API
## Scope & Context

We are implementing the possibility to connect messaging providers to our app.

## Current behavior

After #2814 we can now retrieve the gmail access token and refresh token and store it in the `connectedAccount` table.

## Expected behavior

We want to be able to fetch emails from the gmail API, store them in the database and display them in the frontend.

## Technical inputs

Gmail API docs: https://developers.google.com/gmail/api/guides?hl=en

- Create MESSAGING_PROVIDER_GMAIL_API_KEY env variable and paste the API key from the google cloud console
- Get emails from the API and store them in the database https://developers.google.com/gmail/api/guides/sync?hl=en

### PR DESCRIPTION
**Pull request by bosiraphael**
We are now able to fetch the 500 first messages and threads from the gmail API and store them in the database by using the command `workspace:fetch-messages`:
- We use `users.messages.list` to fetch the messages ids
- We use `users.threads.list` to fetch the thread ids
- We create batch requests to get the emails 100 by 100 (gmail limit).
- We save the threads in the `messageThread` table
- We save the messages in the `message` table

This is a first draft, the code quality is not great yet, but we will add typing and error handling next week + clean the code.

### CODE REVIEW

File: packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    await this.saveMessages(
+      messagesResponse,
+      dataSourceMetadata,
+      workspaceDataSource,
+    );
+
+    return messages;
+  }
+
+  async getGmailClient(refreshToken) {
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts:131]:
```suggestion
  async getGmailClient(refreshToken: string): Promise<any>
```

File: packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    });
+  }
+
+  async saveMessageThreads(
+    threads,
+    dataSourceMetadata,
+    workspaceDataSource,
+    connectedAccountId,
+  ) {
+    const messageChannel = await workspaceDataSource?.query(
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts:158]: messageChannel could be empty

File: packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  async saveMessages(messages, dataSourceMetadata, workspaceDataSource) {
+    for (const message of messages) {
+      const {
+        externalId,
+        headerMessageId,
+        subject,
+        messageThreadId,
+        from,
+        body,
+      } = message;
```
Comment (Weiko) [inline packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts:180]: Not sure why we need to deconstruct here? Not a strong opinion about it, just curious!

**Review by Weiko — APPROVED** [review]
LGTM overall, we should add more typing + checks (empty responses from the DB, existing threads that we don't want to re-insert)
Good work for the poc 👏

File: packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+      `SELECT * FROM ${dataSourceMetadata.schema}."connectedAccount" WHERE "provider" = 'gmail' AND "accountOwnerId" = $1`,
+      [workspaceMemberId],
+    );
+
+    const refreshToken = connectedAccount[0].refreshToken;
+
+    const gmail = await this.getGmailClient(refreshToken);
+
+    const threads = await gmail.users.threads.list({
+      userId: 'me',
```
[the remark this answers is not in the document]
Comment (bosiraphael) [reply packages/twenty-server/src/workspace/messaging/services/fetch-workspace-messages.service.ts:57]: Yes, it fetches the threads from the connected user

**Pull request merged**